x/autopilot/keeper/redeem_stake_test.go

```diff
@@ -0,0 +1,430 @@
+	stakeibctypes "github.com/Stride-Labs/stride/v16/x/stakeibc/types"
+)
+
+func getRedeemStakeStakeibcPacketMetadata(strideAddress, redemptionReceiver string) string {
+	return fmt.Sprintf(`
+		{
+			"autopilot": {
+				"receiver": "%[1]s",
+				"stakeibc": { "action": "RedeemStake", "ibc_receiver": "%[2]s" } 
+			}
+		}`, strideAddress, redemptionReceiver)
+}
+
+// Helper function to mock out all the state needed to test redeem stake
+// The state is mocked out with an atom host zone
+func (s *KeeperTestSuite) SetupAutopilotRedeemStake(featureEnabled bool, redeemAmount sdkmath.Int, depositAddress, redeemerOnStride sdk.AccAddress) {
+	// Set whether the feature is active
+	params := s.App.AutopilotKeeper.GetParams(s.Ctx)
+	params.StakeibcActive = featureEnabled
+	s.App.AutopilotKeeper.SetParams(s.Ctx, params)
+
+	// set epoch tracker to look up epoch unbonding record
+	s.App.StakeibcKeeper.SetEpochTracker(s.Ctx, stakeibctypes.EpochTracker{
+		EpochIdentifier: epochtypes.DAY_EPOCH,
+		EpochNumber:     1,
+	})
+
+	// set epoch unbonding record which will store the new user redemption record
+	s.App.RecordsKeeper.SetEpochUnbondingRecord(s.Ctx, recordstypes.EpochUnbondingRecord{
+		EpochNumber: 1,
+		HostZoneUnbondings: []*recordstypes.HostZoneUnbonding{
+			{
+				HostZoneId:            HostChainId,
+				UserRedemptionRecords: []string{},
+				NativeTokenAmount:     sdk.NewInt(1000000),
+			},
+		},
+	})
+
+	// store the host zone
+	s.App.StakeibcKeeper.SetHostZone(s.Ctx, stakeibctypes.HostZone{
+		ChainId:          HostChainId,
+		Bech32Prefix:     HostBechPrefix, // required to validate claim receiver
+		HostDenom:        HostDenom,
+		RedemptionRate:   sdk.NewDec(1), // used to determine native token amount
+		DepositAddress:   depositAddress.String(),
+		TotalDelegations: redeemAmount, // there must be enough stake to cover the redemption
+	})
+
+	// fund the user with sttokens so they can redeem
+	// (the function being tested is invoked downstream of the IBC transfer)
+	s.FundAccount(redeemerOnStride, sdk.NewCoin("st"+HostDenom, redeemAmount))
+}
+
+// Helper function to confirm that an autopilot redemption succeeded by confirming a redemption
+// record was created and tokens were escrowed in the deposit account
+func (s *KeeperTestSuite) CheckRedeemStakeSucceeded(redeemAmount sdkmath.Int, redeemDenom string, depositAddress sdk.AccAddress) {
+	// check if redeem record is created
+	hostZoneUnbonding, found := s.App.RecordsKeeper.GetHostZoneUnbondingByChainId(s.Ctx, 1, HostChainId)
+	s.Require().True(found)
+	s.Require().True(len(hostZoneUnbonding.UserRedemptionRecords) > 0,
+		"user redemption record should have been created")
+
+	// check that tokens were escrowed
+	escrowBalance := s.App.BankKeeper.GetBalance(s.Ctx, depositAddress, redeemDenom)
+	s.Require().Equal(redeemAmount.Int64(), escrowBalance.Amount.Int64(), "tokens should have been escrowed")
+}
+
+func (s *KeeperTestSuite) TestTryRedeemStake() {
```

Comment:
Should we add a test that tries to redeem twice by the same user in the same day? My sense is we want high confidence that case will work, and might be best to test it with a unit test vs integration test.

Reply:
There's a unit test for this in the PR that allows multiple redemptions per epoch! [[ref](https://github.com/Stride-Labs/stride/blob/ef7d0776a1ff04abff686a8de9c5758627a75100/x/stakeibc/keeper/msg_server_redeem_stake_test.go#L107-L113)]